packages/discovery-provider/src/tasks/index_payment_router.py

```diff
@@ -0,0 +1,858 @@
+                        if exists:
+                            intersection_found = True
+                            break
+                        # Ensure this transaction is still processed
+                        transaction_signature_batch.append(tx_sig)
+
+                # Restart processing at the end of this transaction signature batch
+                last_tx_signature = transactions_array[-1].signature
+
+                # Append batch of processed signatures
+                if transaction_signature_batch:
+                    transaction_signatures.append(transaction_signature_batch)
+
+                # Ensure processing does not grow unbounded
+                if len(transaction_signatures) > TX_SIGNATURES_MAX_BATCHES:
+                    prev_len = len(transaction_signatures)
+                    # Only take the oldest transaction from the transaction_signatures array
+                    # transaction_signatures is sorted from newest to oldest
+                    transaction_signatures = transaction_signatures[
+                        -TX_SIGNATURES_RESIZE_LENGTH:
+                    ]
+                    logger.info(
+                        f"index_payment_router.py | sliced tx_sigs from {prev_len} to {len(transaction_signatures)} entries"
+                    )
+
+        # Reverse batches aggregated so oldest transactions are processed first
+        transaction_signatures.reverse()
+
+        last_tx_sig: Optional[str] = None
+        last_tx = None
+        if transaction_signatures and transaction_signatures[-1]:
+            last_tx_sig = transaction_signatures[-1][0]
+
+        num_txs_processed = 0
+        for tx_sig_batch in transaction_signatures:
+            logger.info(f"index_payment_router.py | processing {tx_sig_batch}")
+            batch_start_time = time.time()
+
+            tx_infos: List[Tuple[GetTransactionResp, str]] = []
+            with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
+                parse_sol_tx_futures = {
+                    executor.submit(
+                        get_sol_tx_info,
+                        solana_client_manager,
+                        str(tx_sig),
+                    ): tx_sig
+                    for tx_sig in tx_sig_batch
+                }
+                for future in concurrent.futures.as_completed(parse_sol_tx_futures):
+                    try:
+                        tx_info = future.result()
+                        if not tx_info:
+                            continue
+                        tx_infos.append(tx_info)
+                    except Exception as exc:
+                        logger.error(
+                            f"index_payment_router.py | error {exc}", exc_info=True
+                        )
+                        raise
+
+            # Sort by slot
+            # Note: while it's possible (even likely) to have multiple tx in the same slot,
```

Comment:
I like the idea of enforcing determinism here (I know you just copied this code) so discovery nodes are consistent between eachother